### PR TITLE
Implement associative scan

### LINKED ISSUE #2708: Implement associative scan
A minimum prerequisite for decent performance on SSM-like models and generalized (cumulative) reduce.

https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.associative_scan.html

### PR DESCRIPTION
**Pull request by blueridanus**
Closes #2708.

Uses [this](https://en.wikipedia.org/wiki/Prefix_sum#Algorithm_1:_Shorter_span,_more_parallel) algo.

- [x] first impl
- [x] support reverse
- [x] replace cumsum (existing cumsum tests will also cover it for free)
- [ ] remove realizes?
- [ ] basic testing
- [ ] axis tests
- [ ] test associativity not violated (i.e. not presuming commutativity)
- [ ] SSM tests
- [ ] some benchmarking
- [ ] compare perf against old cumsum
- [ ] jax comparisons?

### CODE REVIEW

**Comment by blueridanus** [comment]
Works, but awfully slow for some reason.

**Comment by geohot** [comment]
I like this feature, but this hasn't been updated for a while
